tedana/utils/masking.py

```diff
@@ -0,0 +1,125 @@
+"""
+"""
+import numpy as np
+import nibabel as nib
+
+
+def apply_mask_nii_me(imgs, mask_img):
+    """
+    Apply multi-echo mask to set of nifti images. Mask may vary by echo.
+    """
+    pass
+
+
+def apply_mask_gii_me(img, mask_img):
+    """
+    Apply multi-echo mask to set of gifti images. Mask may vary by echo.
+    TODO: Extend to work with multiple images.
+
+    Parameters
+    ----------
+    img : :obj:`nibabel.gifti.gifti.GiftiImage`
+        3D (Vertex, Echo, Time) gifti image containing data.
+    mask_img : :obj:`nibabel.gifti.gifti.GiftiImage`
+        1D (Vertex) or 2D (Vertex, Echo) gifti image with boolean or bool-like
+        values.
+
+    Returns
+    -------
+    masked_arr : :obj:`numpy.ndarray`
+        M x E x T data.
+    """
+    if not isinstance(mask_img, nib.gifti.GiftiImage):
+        raise TypeError('Provided mask is not a GiftiImage.')
+    elif len(mask_img.darrays) > 1:
+        raise ValueError('Mask may only contain one GiftiDataArray.')
+
+    if not isinstance(img, nib.gifti.GiftiImage):
+        raise TypeError('Provided data img is not a GiftiImage.')
+
+    img_arrs = [np.array(darr.data) for darr in img.darrays]
+    all_imgs_same = np.all(arr.shape == img_arrs[0].shape for arr in img_arrs)
+    if not all_imgs_same:
+        raise ValueError('All volumes in data img must have same number of '
+                         'vertices.')
+    n_echos = img_arrs[0].shape[1]
+    n_vols = len(img_arrs)
+
+    mask_arr = np.array(mask_img.darrays[0].data).astype(bool)
+    if img_arrs[0].shape[0] != mask_arr.shape[0]:
+        raise ValueError('Multi-echo mask img and data img must have '
+                         'same number of vertices.')
+
+    if mask_arr.ndim == 1:
+        mask_arr = np.tile(mask_arr, (n_echos, 1))
+    elif mask_arr.shape[1] != img_arrs[0].shape[1]:
+        raise ValueError('Multi-echo mask img and data img must have '
+                         'same number of echos.')
+
+    union_mask = np.any(mask_arr, axis=0)
+    union_idx = np.where(union_mask)[0]
+    masked_arrs = [np.empty((np.sum(union_mask), n_echos)) for _ in
+                   range(len(img_arrs))]
+    for i_vol in range(n_vols):
+        masked_arrs[i_vol][:] = np.nan
```

Comment:
I'm not totally clear on why we want `np.nan` in the masked data. Rather, I understand why the `np.nan` are there, but I think this goes back to my comment above: I'm not sure that we need echo-specific masks! I'm happy to be convinced of their utility, but I can't envision a use case at the current juncture.

Reply:
I guess that having echo-specific masks is the same as using [`make_adaptive_mask`](https://tedana.readthedocs.io/en/latest/generated/tedana.utils.utils.make_adaptive_mask.html#tedana.utils.utils.make_adaptive_mask), but this way only data need to be passed around after initial masking, rather than data + a mask. I'd rather avoid having any masks or mask-like arrays passed around if the data can carry the same information implicitly. It should be easy enough to get the `min_mask` or the `adaptive_mask` from an array that has `NaNs` in voxels from echoes with dropout in those areas.